Initial pass of `libctru` version checking in `ctru-sys`

ian-h-chamberlain opened the pull request:
Partially addresses #100, but we'd probably want to do something as part of the CI release process as well.

This isn't set in stone, I'm just experimenting with some different things to see what makes sense. So far this seems to work pretty well, but I'm not sure if we want to encode the "duplicate" info with the `+` metadata on the build version as well as with the major/minor scheme mentioned in #100 

Let me know what you think, I'll leave as a draft for now and we can discuss it.

ian-h-chamberlain committed:
Initial pass of version checking in ctru-sys

ian-h-chamberlain committed:
Fix ctru dependency version number

AzureMarker commented on ctru-sys/build.rs:
```diff
@@ -14,4 +16,69 @@ fn main() {
+            println!("cargo:rustc-env=LIBCTRU_PATCH={patch}");
+
+            // It would be nice if we could rerun-if-changed on libctru itself,
+            // maybe we can write a file to OUT_DIR and check that file?
```
I don't think that will work since we're the ones changing the file, so it won't change if libctru changes. Instead, we could do a `rerun-if-changed` on the `libctru` static library file. We'd just have to resolve that path.

ian-h-chamberlain replied:
Good point, we already know the library name / search dirs we're linking against so it should be possible to figure out the filename of the actual library being used, which would change if the libs got updated. Ideally we could detect header changes too, but maybe just library is enough...

I suppose in theory we could also take the output of `[dkp-]pacman -Ql libctru` to list _all_ files owned by the package, but I'm not sure if that provides any significant advantage over just using the library. I guess it would include docstring changes ?

ian-h-chamberlain committed:
Output version number in bindgen.sh

Also bump doxgyen-rs to crates.io version, and mark bindings.rs as
generated for Github.

ian-h-chamberlain committed:
Add all `libctru` files to rerun-if-changed

Also simplify version checking logic a bit. It's just a warning, so we
can check for exact version just to be safe and skip the major/minor
comparisons for semver.

Also update the README a bit to explain the versioning scheme.

ian-h-chamberlain committed:
Fix command check for `pacman`

Meziu commented:
We should let people the have the choice to use an external `libctru` binary rather than requiring `dkp-pacman`. It would be helpful to ensure compatibility in the future.

ian-h-chamberlain commented:
> We should let people the have the choice to use an external `libctru` binary rather than requiring `dkp-pacman`. It would be helpful to ensure compatibility in the future.

That's a fair point. This change doesn't prevent anything like that, it would just display a warning that we couldn't determine the `libctru` version (and even then, that would only be displayed if using `-vv` or the compilation/linking failed). Hopefully that warning would give the user a clue that they might need to upgrade or use a supported version instead.

Meziu reviewed (approved):
LGTM 👍
